### PR TITLE
feat(gpu): implement PBSs that use new Hopper's features

### PR DESCRIPTION
**Pull request by pdroalves**
https://github.com/zama-ai/tfhe-rs-internal/issues/450
https://github.com/zama-ai/concrete-internal/issues/345

closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap.h [hunk 1]
```diff
@@ -213,6 +240,46 @@ template <typename Torus> struct pbs_buffer<Torus, PBS_TYPE::CLASSICAL> {
...
+        if (supports_dsm)
+          minimum_sm_tbc =
+              get_buffer_size_sm_dsm_plus_tbc_classic_programmable_bootstrap<
+                  Torus>(polynomial_size);
+
+        uint64_t partial_dm = full_sm - partial_sm;
+        uint64_t full_dm = full_sm;
+        uint64_t device_mem = 0;
+
+        if (max_shared_memory < partial_sm + minimum_sm_tbc) {
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap.h:266]: I don't understand why you have to add partial_sm with minimum_sm_tbc in this check (same for the check below), why is it done like this?

File: backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap_multibit.h [hunk 2]
```diff
@@ -113,12 +144,25 @@ template <typename Torus>
...
+__host__ __device__ uint64_t
+get_buffer_size_full_sm_tbc_multibit_programmable_bootstrap(
+    uint32_t polynomial_size);
 
 template <typename Torus> struct pbs_buffer<Torus, PBS_TYPE::MULTI_BIT> {
   int8_t *d_mem_keybundle = NULL;
   int8_t *d_mem_acc_step_one = NULL;
   int8_t *d_mem_acc_step_two = NULL;
   int8_t *d_mem_acc_cg = NULL;
+  int8_t *d_mem_acc_tbc = NULL;
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap_multibit.h:165]: Maybe add `#if CUDA_ARCH >= 900` here as well

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap.cuh [hunk 3]
```diff
@@ -1,8 +1,120 @@
...
+
+#include "device.h"
+#include "fft/bnsmfft.cuh"
+#include "programmable_bootstrap.h"
+#include "programmable_bootstrap_multibit.h"
+
+#include "cooperative_groups.h"
+
+using namespace cooperative_groups;
+namespace cg = cooperative_groups;
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap.cuh:12]: Maybe these cooperative groups things should be placed in the pbs cg files? They are still there as well btw.

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_classic.cuh [hunk 4]
```diff
@@ -7,112 +7,20 @@
...
 #include "fft/twiddles.cuh"
 #include "polynomial/parameters.cuh"
 #include "polynomial/polynomial_math.cuh"
+#include "programmable_bootstrap.cuh"
 #include "programmable_bootstrap.h"
 #include "types/complex/operations.cuh"
 
-// Cooperative groups are used for this implementation
 using namespace cooperative_groups;
 namespace cg = cooperative_groups;
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_classic.cuh:22]: Here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu [hunk 5]
```diff
@@ -546,3 +773,41 @@ template void scratch_cuda_programmable_bootstrap<uint32_t, int32_t>(
     uint32_t glwe_dimension, uint32_t polynomial_size, uint32_t level_count,
     uint32_t input_lwe_ciphertext_count, uint32_t max_shared_memory,
     bool allocate_gpu_memory);
+
+template bool has_support_to_cuda_programmable_bootstrap_tbc<uint32_t>(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:777]: Missing an #if CUDA_ARCH >= 900 maybe?

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_classic.cuh [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    return cuda_check_support_thread_block_clusters();
+  }
+}
+
+template <typename Torus, class params>
+__host__ bool supports_thread_block_clusters_on_classic_programmable_bootstrap(
+    uint32_t num_samples, uint32_t glwe_dimension, uint32_t polynomial_size,
+    uint32_t level_count, uint32_t max_shared_memory) {
+
+  if (!cuda_check_support_thread_block_clusters() || num_samples > 128)
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_classic.cuh:370]: Why this value, 128?

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_multibit.cuh [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+    return cuda_check_support_thread_block_clusters();
+  }
+}
+
+template <typename Torus, class params>
+__host__ bool supports_thread_block_clusters_on_multibit_programmable_bootstrap(
+    uint32_t num_samples, uint32_t glwe_dimension, uint32_t polynomial_size,
+    uint32_t level_count, uint32_t max_shared_memory) {
+
+  if (!cuda_check_support_thread_block_clusters())
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_multibit.cuh:408]: Here you don't have the same condition on number of inputs

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_multibit.cuh [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+        &config));
+  }
+
+  return cluster_size >= level_count * (glwe_dimension + 1);
+}
+
+template __host__ bool
+supports_distributed_shared_memory_on_multibit_programmable_bootstrap<uint64_t>(
+    uint32_t polynomial_size, uint32_t max_shared_memory);
+#endif // FASTMULTIBIT_PBS_H
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_tbc_multibit.cuh:477]: TBC_MULTIBIT_PBS_H

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu [hunk 9]
```diff
@@ -143,9 +316,19 @@ void scratch_cuda_programmable_bootstrap_32(
...
+          level_count, max_shared_memory))
+    scratch_cuda_programmable_bootstrap_tbc<uint32_t, int32_t>(
+        stream, (pbs_buffer<uint32_t, CLASSICAL> **)buffer, glwe_dimension,
+        polynomial_size, level_count, input_lwe_ciphertext_count,
+        max_shared_memory, allocate_gpu_memory);
+  else
+#endif
+      if (has_support_to_cuda_programmable_bootstrap_cg<uint32_t>(
+              glwe_dimension, polynomial_size, level_count,
+              input_lwe_ciphertext_count, max_shared_memory))
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:331]: Formatting seems a bit off here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu [hunk 10]
```diff
@@ -168,9 +351,19 @@ void scratch_cuda_programmable_bootstrap_64(
...
+  if (has_support_to_cuda_programmable_bootstrap_tbc<uint64_t>(
+          input_lwe_ciphertext_count, glwe_dimension, polynomial_size,
+          level_count, max_shared_memory))
+    scratch_cuda_programmable_bootstrap_tbc<uint64_t, int64_t>(
+        stream, (pbs_buffer<uint64_t, CLASSICAL> **)buffer, glwe_dimension,
+        polynomial_size, level_count, input_lwe_ciphertext_count,
+        max_shared_memory, allocate_gpu_memory);
+  else
+#endif
+      if (has_support_to_cuda_programmable_bootstrap_cg<uint64_t>(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:364]: Same here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu [hunk 11]
```diff
@@ -330,9 +523,26 @@ void cuda_programmable_bootstrap_lwe_ciphertext_vector_32(
...
+        static_cast<uint32_t *>(lut_vector_indexes),
+        static_cast<uint32_t *>(lwe_array_in),
+        static_cast<uint32_t *>(lwe_input_indexes),
+        static_cast<double2 *>(bootstrapping_key),
+        (pbs_buffer<uint32_t, CLASSICAL> *)buffer, lwe_dimension,
+        glwe_dimension, polynomial_size, base_log, level_count, num_samples,
+        num_luts, lwe_idx, max_shared_memory);
+  else
+#endif
+      if (has_support_to_cuda_programmable_bootstrap_cg<uint32_t>(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:543]: Same here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu [hunk 12]
```diff
@@ -441,9 +651,26 @@ void cuda_programmable_bootstrap_lwe_ciphertext_vector_64(
...
+        static_cast<uint64_t *>(lut_vector_indexes),
+        static_cast<uint64_t *>(lwe_array_in),
+        static_cast<uint64_t *>(lwe_input_indexes),
+        static_cast<double2 *>(bootstrapping_key),
+        (pbs_buffer<uint64_t, CLASSICAL> *)buffer, lwe_dimension,
+        glwe_dimension, polynomial_size, base_log, level_count, num_samples,
+        num_luts, lwe_idx, max_shared_memory);
+  else
+#endif
+      if (has_support_to_cuda_programmable_bootstrap_cg<uint64_t>(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:671]: Again

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cu [hunk 13]
```diff
@@ -191,9 +246,26 @@ void cuda_multi_bit_programmable_bootstrap_lwe_ciphertext_vector_64(
...
+        static_cast<uint64_t *>(lut_vector_indexes),
+        static_cast<uint64_t *>(lwe_array_in),
+        static_cast<uint64_t *>(lwe_input_indexes),
+        static_cast<uint64_t *>(bootstrapping_key),
+        (pbs_buffer<uint64_t, MULTI_BIT> *)buffer, lwe_dimension,
+        glwe_dimension, polynomial_size, grouping_factor, base_log, level_count,
+        num_samples, num_luts, lwe_idx, max_shared_memory, lwe_chunk_size);
+  else
+#endif
+      if (supports_cooperative_groups_on_multibit_programmable_bootstrap<
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cu:266]: Same format thing in this file. I guess if the formatting tool doesn't complain we could leave it as is.

**Review by agnesLeroy — COMMENTED** [review]
Hi @pdroalves! Thanks a lot for this! I haven't looked in details at the tbc PBS code itself, would be way too long for me to do so, but here's a my review on the rest.

File: backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap_multibit.h (hunk 2 above)
Comment (pdroalves) [reply backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap_multibit.h:165]: Good one!

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu (hunk 10 above)
Comment (pdroalves) [reply backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:364]: The first part of the conditional is within the `#if` conditional. clang is aligning this if with the else at line 327. If I change that `make fmt_gpu` will change back.

**Comment by pdroalves** [comment]
Since TBC PBS will only run (and be compiled) on H100s, I rebased this branch on top of `gpu/hyperstack_ci` to run on Hyperstack, as suggested by @david-zama .

**Comment by pdroalves** [comment]
@slab-ci gpu_test

**Comment by pdroalves** [comment]
@agnesLeroy Looks like this breaks the CI. Should I undo this rebase?

**Comment by agnesLeroy** [comment]
@pdroalves you should target main instead of David's branch

**Comment by pdroalves** [comment]
Indeed. 

I rebased on `chre/gpu/integer_mult_gpu_params`, so we have a good chain of PRs to merge. This would be the first.

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu (hunk 9 above)
Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_classic.cu:331]: Ok makes sensse :+1:

File: backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap.h (hunk 1 above)
Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/include/programmable_bootstrap.h:266]: Ok makes sense, could you maybe add this explanation as a comment here please? Thanks! :pray:

**Comment by pdroalves** [comment]
@agnesLeroy  This branch passes integer tests on H100.

**Pull request merged**